### ISSUE TITLE
allow string literals (unions of string literals?) as a type of an indexer

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#7656]
```
type Keyed<a> = { [key: 'this' | 'that']: a; } // expected to work, actual doesn't work
```

**Comment by malibuzios** [comment]
I also encountered this and considered opening an issue, though having a string literal union as a key type would have very similar semantics to:
```ts
type Keyed<a> = { 'this': a; 'that': a }
```
Which would appear a bit more verbose and require more typing though.

However, the indexer approach does have the advantage that it allows aliasing the key type (the string literal union) separately, e.g:
```ts
type RequestMethod = "GET" | "SET" | "PUT" | "UPDATE" | "DELETE";
type handlers = { [key: RequestMethod]: (requestObject: RequestObject) => boolean; };
```
I think that based on the usefulness of this particular pattern it should definitely be considered to become valid syntax.

Also, the same pattern may also prove useful when the [numeric literal type](https://github.com/Microsoft/TypeScript/pull/7480) is introduced:

```ts
type ValidNumericKey = 1 | 2 | 4 | 8 | 16;
type values = { [key: ValidNumericKey]: SomeObject };
```

**Comment by aleksey-bykov** [comment]
it's not an equivalent, my Keyed can take from 0 through 1 to 2 values,
whereas yours always needs exactly 2

completely different semantics

**Comment by malibuzios** [comment]
I was just about to fix it when you replied. I changed it to make all properties optional. Though it may still not have the same exact semantics. Can you think of any other differences?

**Comment by malibuzios** [comment]
Well there's one difference (that's isn't really about semantics) that the interface style notation, based on current behavior, would allow using the dot (`.`) operator, but the indexer wouldn't:
```ts
type Keyed1<a> = { this?: a; that?: a }
type Keyed2<a> = { [key: 'this' | 'that']: a; }

let x: Keyed1<number> = {};
let y: Keyed2<number> = {};

x['that'] = 34; // OK
x.that = 34; // OK

y['that'] = 34 // OK
y.that = 34 // Error
```
However that behavior could be changed in case the key is a string literal, I guess, perhaps, maybe?

**Comment by RyanCavanaugh** [comment]
Can you explain what you expect this to do?

Is it equivalent to this?
```ts
type Keyed<a> = { 'this': a } | { 'that': a };
```

**Comment by malibuzios** [comment]
@RyanCavanaugh 

I believe (though of course I may be subtly wrong) the idea is that the string literal union constrains what keys can be used to a finite set, in this case `{"this", "that"}`. For simplicity let's assume `a` is a `number`

```ts
type Keyed = { [key: 'this' | 'that']: number; }
```
I believe the possible values for `let x: Keyed` are:
```ts
x = {};
// or
x = { 'this': ...some number... }
// or
x = { 'that': ...some number... }
//or 
x = { 'this': ...some number...,  'that': ...some number... }
```

**Comment by RyanCavanaugh** [comment]
So `{ 'this': 4, 'something': 6 }` would be invalid?

**Comment by malibuzios** [comment]
@RyanCavanaugh 

Based on my understanding that would seem like the most natural interpretation of this combination of the indexer syntax and the concept of string literal union. 

It also seems like this may mean that unlike the semantics of an "normal" interface this would be a strict type, i.e. not assignable from a wider type (having additional properties)?

**Comment by malibuzios** [comment]
@RyanCavanaugh 
Or the fact it's still enclosed within the "brackets" `{}`, representing an interface, means it's still extensible?

**Comment by aleksey-bykov** [comment]
why? to limit the possible set of keys, prevent unexpected keys from
possibly being used

**Comment by malibuzios** [comment]
@RyanCavanaugh 

I'm sort of rethinking this. Perhaps we need to separate the two different aspects of semantics here:

1. What can be indexed by this - Which based on the natural interpretation of literal types I believe should only include the string "this", "that" and no more.
2. What can be assigned to this - Is this a "strict" interface that only takes the 4 variations I listed, or does this behave like an all optional interface? which very unfortunately today is even [assignable by anything](https://github.com/Microsoft/TypeScript/pull/3842) -  personally I feel this is one of those places where I'm amazed still happens and nothing has really been done about.

**Comment by malibuzios** [comment]
@aleksey-bykov 

I think we have a bigger problem here: it seems like indexer key types are not even checked for conventional types like `string` and `number` (there may be an open issue for this but I haven't searched yet):

These two examples doesn't induce any no errors whatsoever:
```ts
type A = {[key: number]: any};

let a: A = {};
a["abcd"] = 1; // No error?
let x =  a["abcd"]; // No error?
```

```ts
type B = {[key: string]: any};

let b: B = {};
b[123] = 1; // No error?
let y = b[123]; // No error?
```

**Comment by malibuzios** [comment]
This issue is a duplicate. This conversation needs to be somehow merged with #5683 which seems to be the main tracking point this subject (there has been several duplicates so far).

I opened a different issue (#7660) for the indexer key type checking.

**Comment by malibuzios** [comment]
@aleksey-bykov @RyanCavanaugh 

[This comment](https://github.com/Microsoft/TypeScript/issues/7660#issuecomment-200786716) in #7660 is a direct continuation of the discussion here, but applied to the more general issue of the how strictly should index signature keys be type-checked.
